src/scripts/services/group-service.ts

```diff
@@ -30,21 +36,123 @@ export const oktaGroupAsGroup = (oktaGroup: okta.Group): Group => ({
+          .then(O.some)
+          .catch((error) => {
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+            return typeof error === 'object' && error.status === 404
+              ? Promise.resolve(O.none)
+              : Promise.reject(error);
+          }),
+      (error: unknown) =>
+        `Failed fetching group details for [${groupId}] because of [${JSON.stringify(
+          error
+        )}]`
+    );
+
+  readonly addUserToGroup = (
+    groupId: string,
+    userId: string
+  ): TE.TaskEither<string, Response> =>
+    TE.tryCatch(
+      // eslint-disable-next-line functional/functional-parameters
+      () =>
+        // eslint-disable-next-line functional/no-this-expression
+        this.client
+          .getUser(userId)
+          // eslint-disable-next-line
+          .then((user: okta.User) => user.addToGroup(groupId)),
+      (error: unknown) =>
+        `Failed to add user [${userId}] to group [${groupId}] because of [${JSON.stringify(
+          error
+        )}].`
+    );
+
+  readonly removeUserFromGroup = (
+    groupId: string,
+    userId: string
+  ): TE.TaskEither<string, Response> =>
+    TE.tryCatch(
+      // eslint-disable-next-line functional/functional-parameters
+      () =>
+        // eslint-disable-next-line functional/no-this-expression
+        this.client
+          .getGroup(groupId)
+          // eslint-disable-next-line
+            .then((group: okta.Group) => group.removeUser(userId)),
+      (error: unknown) =>
+        `Failed to remove user [${userId}] to group [${groupId}] because of [${JSON.stringify(
+          error
+        )}].`
+    );
+
+  // eslint-disable-next-line functional/functional-parameters
+  readonly listGroups = (): TE.TaskEither<string, readonly Group[]> => {
+    // We need to populate groups with all of the client data so it can be
+    // returned. Okta's listGroups() function returns a custom collection that
+    // does not allow for any form of mapping, so array mutation is needed.
+
+    return TE.tryCatch(
+      // eslint-disable-next-line functional/functional-parameters
+      () => {
+        // eslint-disable-next-line functional/prefer-readonly-type
+        const groups: Group[] = [];
+
+        return (
+          // eslint-disable-next-line functional/no-this-expression
+          this.client
+            .listGroups()
+            // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
+            .each((oktaGroup) => {
+              // eslint-disable-next-line functional/immutable-data
+              return groups.push(oktaGroupAsGroup(oktaGroup));
+            })
+            // eslint-disable-next-line functional/functional-parameters
+            .then(() => {
+              return groups;
+            })
+        );
+      },
+      (error: unknown) =>
+        `Failed to list groups because of [${JSON.stringify(error)}].`
+    );
+  };
+
+  readonly validateGroupExists = (
```

Comment:
Please move this somewhere other than the service:
1. Service should generally be un-opinionated about the meaning of existence.
2. It doesn't rely on the Okta API, but can (as it currently stands) only be tested when you instantiate the service.

The same applies to the equivalent method in UserService.

Reply:
Perhaps create a ts file called `validation-service`?

Reply:
Sure. This method has been overused though. You should only use it when you want this particular error shape.

Reply:
added to new file